invertedai/large/drive.py

```diff
@@ -0,0 +1,188 @@
+from invertedai.api.drive import DriveResponse
+from ._quadtree import QuadTreeAgentInfo, QuadTree, _flatten_and_sort, QUADTREE_SIZE_BUFFER
+
+async def async_drive_all(async_input_params):
+    all_responses = await asyncio.gather(*[iai.async_drive(**input_params) for input_params in async_input_params])
+    return all_responses
+
+@validate_call
+def large_drive(
+    location: str,
+    agent_states: List[AgentState],
+    agent_attributes: List[AgentAttributes],
+    recurrent_states: List[RecurrentState],
+    traffic_lights_states: Optional[TrafficLightStatesDict] = None,
+    light_recurrent_states: Optional[List[LightRecurrentState]] = None,
+    get_infractions: bool = False,
+    random_seed: Optional[int] = None,
+    api_model_version: Optional[str] = None,
+    single_call_agent_limit: Optional[int] = None,
+    async_api_calls: bool = True
+) -> DriveResponse:
+    """
+    A utility function to drive more than the normal capacity of agents in a call to :func:`drive`.
+    The agents are inserted into a quadtree structure and :func:`drive` is then called on each
+    region represented by a leaf node of the quadtree. Agents near this region are included in the
+    :func:`drive` calls to ensure the agents behave properly locally. The quadtree is constructed 
+    during each call to this utility function to maintain statelessness.
+
+    Parameters
+    ----------
+    location:
+        Location name in IAI format.
+
+    agent_states:
+        Current states of all agents.
+        The state must include x: [float], y: [float] coordinate in meters
+        orientation: [float] in radians with 0 pointing along x and pi/2 pointing along y and
+        speed: [float] in m/s.
+
+    agent_attributes:
+        Static attributes of all agents.
+        List of agent attributes. Each agent requires, length: [float]
+        width: [float] and rear_axis_offset: [float] all in meters. agent_type: [str],
+        currently supports 'car' and 'pedestrian'.
+        waypoint: optional [Point], the target waypoint of the agent.
+
+    recurrent_states:
+        Recurrent states for all agents, obtained from the previous call to
+        :func:`drive` or :func:`initialize` or the large equivalents.
+
+    traffic_lights_states:
+       If the location contains traffic lights within the supported area,
+       their current state should be provided here. Any traffic light for which no
+       state is provided will have a state generated by iai.
+
+    light_recurrent_states:
+       Light recurrent states for all agents, obtained from the previous call to
+        :func:`drive` or :func:`initialize` or the large equivalents.
+       Specifies the state and time remaining for each light group in the map.
+       If manual control of individual traffic lights is desired, modify the relevant state(s) 
+       in traffic_lights_states, then pass in light_recurrent_states as usual.
```

Comment:
Consider pointing the reader to the documentation for drive instead. That has its downsides, but makes it less likely that this description gets out of sync from the one in drive.

Reply:
I like this idea, I will do the same with `large_initialize` wherever possible because yeah it might be annoying but it is much more convenient for maintainability and less prone to mistakes and confusion.